From 41e762d00eec30b568a16eaf36deb8eeda86c258 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:53:24 -0400
Subject: [fs] nfs: commit layouts in fdatasync

Message-id: <1403798077-28799-6-git-send-email-steved@redhat.com>
Patchwork-id: 84876
O-Subject: [RHEL7.1 PATCH 05/78] nfs: commit layouts in fdatasync
Bugzilla: 1113248
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Christoph Hellwig <hch@infradead.org>

 "fdatasync() is similar to fsync(), but does not flush modified metadata
  unless that metadata is needed in order  to  allow  a  subsequent  data
  retrieval to be correctly handled."

We absolutely need to commit the layouts to be able to retrieve the data
in case either the client, the server or the storage subsystem go down.

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upsteam Id: 1b33809ea8c671ccbceeaaa8d842631b441bed54
Bugzilla: 1113248
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/nfs4file.c b/fs/nfs/nfs4file.c
index 9c8f09a..cb5f89c 100644
--- a/fs/nfs/nfs4file.c
+++ b/fs/nfs/nfs4file.c
@@ -100,8 +100,7 @@ nfs4_file_fsync(struct file *file, loff_t start, loff_t end, int datasync)
    break;
   mutex_lock(&inode->i_mutex);
   ret = nfs_file_fsync_commit(file, start, end, datasync);
-  if (!ret && !datasync)
-   /* application has asked for meta-data sync */
+  if (!ret)
    ret = pnfs_layoutcommit_inode(inode, true);
   mutex_unlock(&inode->i_mutex);
   /*
-- 
1.7.1